File: test/python3/libdnf5/advisory/test_advisory.py
# Copyright Contributors to the libdnf project.
# SPDX-License-Identifier: GPL-2.0-or-later
#
# This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
#
# Libdnf is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# Libdnf is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with libdnf.  If not, see <https://www.gnu.org/licenses/>.

import libdnf5

import base_test_case


class TestAdvisory(base_test_case.BaseTestCase):
    def setUp(self):
        super().setUp()
        self.add_repo_repomd("repomd-repo1")
        query = libdnf5.advisory.AdvisoryQuery(self.base)
        query.filter_type("security")
        self.advisory = next(iter(query))

    def test_advisory_query(self):
        query = libdnf5.advisory.AdvisoryQuery(self.base)
        query.filter_name("DNF-2019-1")
        self.assertEqual(query.size(), 1)
        advisory = next(iter(query))
        self.assertEqual(advisory.get_name(), "DNF-2019-1")

        query = libdnf5.advisory.AdvisoryQuery(self.base)
        query.filter_type("bugfix")
        self.assertEqual(query.size(), 2)
        advisory = next(iter(query))
        self.assertEqual(advisory.get_type(), "bugfix")

        query = libdnf5.advisory.AdvisoryQuery(self.base)
        query.filter_reference("1111")
        self.assertEqual(query.size(), 1)
        advisory = next(iter(query))
        refs = advisory.get_references()
        self.assertEqual(len(refs), 1)
        self.assertEqual(refs[0].get_id(), "1111")

        query = libdnf5.advisory.AdvisoryQuery(self.base)
        query.filter_severity("moderate")
        self.assertEqual(query.size(), 1)
        advisory = next(iter(query))
        self.assertEqual(advisory.get_severity(), "moderate")

        query = libdnf5.advisory.AdvisoryQuery(self.base)
        pkgquery = libdnf5.rpm.PackageQuery(self.base)
        query.filter_packages(pkgquery)
        self.assertEqual(query.size(), 1)
        advisory = next(iter(query))
        self.assertEqual(advisory.get_name(), "DNF-2019-1")

        query = libdnf5.advisory.AdvisoryQuery(self.base)
        pkgquery = libdnf5.rpm.PackageQuery(self.base)
        pkgs = query.get_advisory_packages_sorted(pkgquery)
        self.assertEqual(pkgs.size(), 1)
        self.assertEqual(pkgs[0].get_nevra(), "pkg-1.2-3.x86_64")

    def test_advisory(self):
        self.assertEqual(self.advisory.get_name(), "DNF-2019-1")
        self.assertEqual(self.advisory.get_severity(), "moderate")
        self.assertEqual(self.advisory.get_type(), "security")
        self.assertEqual(self.advisory.get_buildtime(), 1550849401)
        self.assertEqual(self.advisory.get_vendor(), "dnf-testing@redhat.com")
        self.assertEqual(self.advisory.get_description(),
                         "testing advisory 2019")
        self.assertEqual(self.advisory.get_title(), "bugfix_A-1.0-1")
        self.assertEqual(self.advisory.get_status(), "stable")
        self.assertEqual(self.advisory.get_rights(), "")
        self.assertEqual(self.advisory.get_message(), "")
        self.assertEqual(self.advisory.is_applicable(), True)

    def test_references(self):
        refs = self.advisory.get_references()
        self.assertEqual(len(refs), 1)
        reference = refs[0]
        self.assertEqual(reference.get_id(), "1111")
        self.assertEqual(reference.get_type(), "cve")
        self.assertEqual(reference.get_title(), "CVE-2999")
        self.assertEqual(reference.get_url(), "https://foobar/foobarupdate_2")

    def test_collections(self):
        cols = self.advisory.get_collections()
        self.assertEqual(len(cols), 1)
        collection = cols[0]
        self.assertEqual(collection.get_advisory_id(), self.advisory.get_id())
        self.assertEqual(collection.get_advisory(), self.advisory)
        self.assertEqual(collection.is_applicable(), True)

    def test_advisory_packages(self):
        collection = self.advisory.get_collections()[0]
        adv_packages = collection.get_packages()
        self.assertEqual(len(adv_packages), 2)
        adv_package = adv_packages[0]

        self.assertEqual(adv_package.get_advisory_id(), self.advisory.get_id())
        self.assertEqual(adv_package.get_advisory(), self.advisory)
        # We can't compare the collections directly because adv_package constructs
        # a new instance of the same collection
        self.assertEqual(adv_package.get_advisory_collection(
        ).get_advisory_id(), self.advisory.get_id())

        self.assertEqual(adv_package.get_name(), "pkg")
        self.assertEqual(adv_package.get_epoch(), "0")
        self.assertEqual(adv_package.get_version(), "1.2")
        self.assertEqual(adv_package.get_release(), "3")
        self.assertEqual(adv_package.get_evr(), "1.2-3")
        self.assertEqual(adv_package.get_arch(), "x86_64")
        self.assertEqual(adv_package.get_nevra(), "pkg-1.2-3.x86_64")

        self.assertEqual(adv_package.get_reboot_suggested(), True)
        self.assertEqual(adv_package.get_restart_suggested(), False)
        self.assertEqual(adv_package.get_relogin_suggested(), False)

    def test_advisory_modules(self):
        collection = self.advisory.get_collections()[0]
        adv_modules = collection.get_modules()
        self.assertEqual(len(adv_modules), 2)
        adv_module = adv_modules[0]

        self.assertEqual(adv_module.get_advisory_id(), self.advisory.get_id())
        self.assertEqual(adv_module.get_advisory(), self.advisory)
        # We can't compare the collections directly because adv_module constructs
        # a new instance of the same collection
        self.assertEqual(adv_module.get_advisory_collection(
        ).get_advisory_id(), self.advisory.get_id())

        self.assertEqual(adv_module.get_name(), "perl-DBI")
        self.assertEqual(adv_module.get_stream(), "master")
        self.assertEqual(adv_module.get_version(), "2")
        self.assertEqual(adv_module.get_context(), "2a")
        self.assertEqual(adv_module.get_arch(), "x86_64")
        self.assertEqual(adv_module.get_nsvca(), "perl-DBI:master:2:2a:x86_64")

File: test/libdnf5/advisory/test_advisory.cpp
// Copyright Contributors to the libdnf project.
// SPDX-License-Identifier: GPL-2.0-or-later
//
// This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
//
// Libdnf is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
//
// Libdnf is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with libdnf.  If not, see <https://www.gnu.org/licenses/>.


#include "test_advisory.hpp"

#include <libdnf5/rpm/package_query.hpp>
#include <libdnf5/rpm/package_set.hpp>

#include <filesystem>
#include <set>
#include <vector>


CPPUNIT_TEST_SUITE_REGISTRATION(AdvisoryAdvisoryTest);

//This allows running only this single test suite, by using `getRegistry("AdvisoryAdvisoryTest_suite")` in run_tests.cpp
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION(AdvisoryAdvisoryTest, "AdvisoryAdvisoryTest_suite");

void AdvisoryAdvisoryTest::setUp() {
    BaseTestCase::setUp();
    BaseTestCase::add_repo_repomd("repomd-repo1");
}

void AdvisoryAdvisoryTest::test_get_name() {
    // Tests get_name method
    libdnf5::advisory::AdvisoryQuery advisories(base);
    advisories.filter_type("security");
    libdnf5::advisory::Advisory advisory = *advisories.begin();
    CPPUNIT_ASSERT_EQUAL(advisory.get_name(), std::string("DNF-2019-1"));
}

void AdvisoryAdvisoryTest::test_get_type() {
    // Tests get_type method
    libdnf5::advisory::AdvisoryQuery advisories(base);
    advisories.filter_type("security");
    libdnf5::advisory::Advisory advisory = *advisories.begin();
    CPPUNIT_ASSERT_EQUAL(advisory.get_type(), std::string("security"));
}

void AdvisoryAdvisoryTest::test_get_severity() {
    // Tests get_severity method
    libdnf5::advisory::AdvisoryQuery advisories(base);
    advisories.filter_type("security");
    libdnf5::advisory::Advisory advisory = *advisories.begin();
    CPPUNIT_ASSERT_EQUAL(advisory.get_severity(), std::string("moderate"));
}

void AdvisoryAdvisoryTest::test_get_references() {
    // Tests get_references method
    libdnf5::advisory::AdvisoryQuery advisories(base);
    advisories.filter_type("security");
    libdnf5::advisory::Advisory advisory = *advisories.begin();
    std::vector<libdnf5::advisory::AdvisoryReference> refs = advisory.get_references();
    CPPUNIT_ASSERT_EQUAL((size_t)1, refs.size());

    libdnf5::advisory::AdvisoryReference r = refs[0];
    CPPUNIT_ASSERT_EQUAL(std::string("1111"), r.get_id());
    CPPUNIT_ASSERT_EQUAL(std::string("cve"), r.get_type());
    CPPUNIT_ASSERT_EQUAL(std::string("CVE-2999"), r.get_title());
    CPPUNIT_ASSERT_EQUAL(std::string("https://foobar/foobarupdate_2"), r.get_url());
}

void AdvisoryAdvisoryTest::test_get_collections() {
    // Tests get_collections method
    libdnf5::advisory::AdvisoryQuery advisories(base);
    advisories.filter_type("security");
    libdnf5::advisory::Advisory advisory = *advisories.begin();
    std::vector<libdnf5::advisory::AdvisoryCollection> colls = advisory.get_collections();
    CPPUNIT_ASSERT_EQUAL((size_t)1, colls.size());

    libdnf5::advisory::AdvisoryCollection c = colls[0];
    std::vector<libdnf5::advisory::AdvisoryPackage> pkgs = c.get_packages();
    CPPUNIT_ASSERT_EQUAL((size_t)2, pkgs.size());
    CPPUNIT_ASSERT_EQUAL(std::string("pkg"), pkgs[0].get_name());
    CPPUNIT_ASSERT_EQUAL(std::string("filesystem"), pkgs[1].get_name());

    std::vector<libdnf5::advisory::AdvisoryModule> mods = c.get_modules();
    CPPUNIT_ASSERT_EQUAL((size_t)2, mods.size());
    CPPUNIT_ASSERT_EQUAL(std::string("perl-DBI"), mods[0].get_name());
    CPPUNIT_ASSERT_EQUAL(std::string("ethereum"), mods[1].get_name());

    libdnf5::advisory::AdvisoryQuery adv_query(base);
    adv_query.filter_name("DNF-2020-1");
    CPPUNIT_ASSERT_EQUAL((size_t)1, adv_query.size());
    colls = (*adv_query.begin()).get_collections();
    CPPUNIT_ASSERT_EQUAL((size_t)2, colls.size());

    libdnf5::advisory::AdvisoryCollection c1 = colls[0];
    pkgs.clear();
    pkgs = c1.get_packages();
    CPPUNIT_ASSERT_EQUAL((size_t)2, pkgs.size());
    CPPUNIT_ASSERT_EQUAL(std::string("wget"), pkgs[0].get_name());
    CPPUNIT_ASSERT_EQUAL(std::string("yum"), pkgs[1].get_name());

    mods.clear();
    mods = c1.get_modules();
    CPPUNIT_ASSERT_EQUAL((size_t)1, mods.size());
    CPPUNIT_ASSERT_EQUAL(std::string("perl-DBI"), mods[0].get_name());

    libdnf5::advisory::AdvisoryCollection c2 = colls[1];
    pkgs.clear();
    pkgs = c2.get_packages();
    CPPUNIT_ASSERT_EQUAL((size_t)1, pkgs.size());
    CPPUNIT_ASSERT_EQUAL(std::string("bitcoin"), pkgs[0].get_name());

    mods.clear();
    mods = c2.get_modules();
    CPPUNIT_ASSERT_EQUAL((size_t)1, mods.size());
    CPPUNIT_ASSERT_EQUAL(std::string("perl"), mods[0].get_name());
}
